File: 프로그래머스/2/131704. 택배상자/택배상자.java
import java.util.*;

public class Solution {
    public int solution(int[] order) {
        int answer = 0;
        Stack<Integer> stack = new Stack<>();
        
        int originCur = 0;
        int orderCur = 0;
        
        while (orderCur < order.length) {
            int orderBox = order[orderCur];
            int originBox = originCur + 1;

            if (orderBox == originBox) {
                answer++;
                originCur++;
                orderCur++;
                continue;
            }
            if (!stack.isEmpty() && orderBox == stack.peek()) {
                stack.pop();
                answer++;
                orderCur++;
                continue;
            }
            
            if (!stack.isEmpty() && stack.peek() > orderBox) {
                return answer;
            } else {
                stack.push(originBox);
                originCur++;
            }
        }
        return answer;
    }
}

File: 프로그래머스/2/131704. 택배상자/택배상자.js
function solution(order) {
    let answer = 0;
    const stack = [];
    
    let originCur = 0;
    let orderCur = 0;
    
    while (orderCur < order.length) {
        const orderBox = order[orderCur]
        const originBox = originCur + 1;

        if (orderBox === originBox) {
            answer++;
            originCur++;
            orderCur++;
            continue;
        }
        if (orderBox === stack.at(-1)) {
            stack.pop();
            answer++;
            orderCur++;
            continue;
        } 
        
        if (stack.at(-1) > orderBox) {
            return answer;
        } else {
            stack.push(originBox)
            originCur++;
        }
    }
    return answer;
}
